src/Identity/Swauth/Service.php

```diff
@@ -42,14 +39,34 @@ public static function factory(ClientInterface $client): self
      */
     public function authenticate(array $values): array
     {
-        $response = $this->execute(self::TOKEN_DEFINITION, array_intersect_key($values, self::TOKEN_DEFINITION['params']));
 
-        $token = $this->model(Token::class, $response);
+        if (!empty($values['cachedToken'])) {
+            $token = $this->generateTokenFromCache($values['cachedToken']);
+
+            if ($token->hasExpired()) {
+                throw new \RuntimeException(sprintf('Cached token has expired on "%s".', $token->expires->format(\DateTime::ISO8601)));
+            }
+            $base_uri = $token->catalog->base_uri;
+        } else {
+            $response = $this->execute(self::TOKEN_DEFINITION, array_intersect_key($values, self::TOKEN_DEFINITION['params']));
+            $token = $this->model(Token::class, $response);
+            $base_uri = ($this->model(ServiceCatalog::class, $response))->base_uri;
+        }
 
-        $serviceCatalog = $this->model(ServiceCatalog::class, $response);
         return [
             $token,
-            $serviceCatalog->base_uri,
+            $base_uri,
         ];
     }
+
+    public function generateTokenFromCache(array $cachedToken): Token
+    {
+        return $this->model(Token::class)->populateFromArray($cachedToken);
+    }
+
+    public function generateToken(array $options = []): Token
```

Comment:
consider to indicate which interface this function implements (e.g. with a phpdoc comment) since it isn't called from within the class.